### PR TITLE
feat: Add web3 indexer

### PR DESCRIPTION
**Pull request by bitrocks**

### CODE REVIEW

File: crates/block-producer/src/main.rs [hunk 1]
```diff
@@ -102,12 +104,37 @@ fn run() -> Result<()> {
...
+    // create web3 indexer
+    let web3_indexer = match config.web3_indexer {
+        Some(web3_indexer_config) => {
+            let pool = smol::block_on(async {
+                PgPoolOptions::new()
+                    .max_connections(5)
+                    .connect(&web3_indexer_config.database_url)
+                    .await
+            })?;
+            let l2_sudt_type_script_hash = web3_indexer_config.l2_sudt_type_script_hash;
```
Comment (jjyr) [inline crates/block-producer/src/main.rs:116]:
```suggestion
            let l2_sudt_type_script_hash = rollup_config.l2_sudt_validator_script_type_hash();
```

File: crates/block-producer/src/main.rs [hunk 2]
```diff
@@ -102,12 +104,37 @@ fn run() -> Result<()> {
...
+            let polyjuce_type_script_hash = web3_indexer_config.polyjuice_script_type_hash;
+            let rollup_type_hash = web3_indexer_config.rollup_type_hash;
```
Comment (jjyr) [inline crates/block-producer/src/main.rs:118]:
```suggestion
            let rollup_type_hash = rollup_script_hash;
```

Comment (jjyr) [reply crates/block-producer/src/main.rs:118]: We can read these two values from the current context.

File: crates/tools/src/main.rs [hunk 3]
```diff
@@ -115,6 +115,13 @@ fn main() {
                         .required(true)
                         .help("Polyjuice binaries directory path"),
                 )
+                .arg(
+                    Arg::with_name("database-url")
+                        .short("d")
+                        .takes_value(true)
+                        .required(true)
```
Comment (jjyr) [inline crates/tools/src/main.rs:122]: Make it optional

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/tools/src/main.rs (hunk 3 above)
Comment (bitrocks) [reply crates/tools/src/main.rs]: updated.

**Review by jjyr — APPROVED** [review]

**Pull request merged**